service/internal/presigned-url/middleware.go

```diff
@@ -3,35 +3,43 @@ package presignedurl
 import (
 	"context"
 	"fmt"
-	"net/http"
 
 	awsmiddleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
+	v4 "github.com/aws/aws-sdk-go-v2/aws/signer/v4"
+
 	"github.com/awslabs/smithy-go/middleware"
 )
 
 // ParameterAccessor provides an collection of accessor to for retrieving and
 // setting the values needed to PresignedURL generation
 type ParameterAccessor struct {
+	// GetPresignedURL accessor points to a function that retrieves a presigned url if present
 	GetPresignedURL func(interface{}) (string, bool, error)
+
+	// GetSourceRegion accessor points to a function that retrieves source region for presigned url
 	GetSourceRegion func(interface{}) (string, bool, error)
 
-	CopyInput            func(interface{}) (interface{}, error)
+	// CopyInput accessor points to a function that takes in an input, and returns a copy.
+	CopyInput func(interface{}) (interface{}, error)
+
+	// SetDestinationRegion accessor points to a function that sets destination region on api input struct
 	SetDestinationRegion func(interface{}, string) error
-	SetPresignedURL      func(interface{}, string) error
-}
 
-// URLPresigner provides the interface to presign the input parameters in to a
-// presigned URL.
-type URLPresigner interface {
-	PresignURL(ctx context.Context, srcRegion string, params interface{}) (
-		presignedURL string, signedHeader http.Header, err error,
-	)
+	// SetPresignedURL accessor points to a function that sets presigned url on api input struct
+	SetPresignedURL func(interface{}, string) error
+
+	// PresignOperation is the presign function accessor
+	PresignOperation func(ctx context.Context, client interface{},
+		srcRegion string, params interface{}) (*v4.PresignedHTTPRequest, error)
```

Comment:
Down side of putting this as a function on the accessory instead of an interface on the middleware's `Options` is it creates the opportunity for a runtime error because types are not matched up correctly.

What do you think about leaving in the unexported type for wrapping a API client's `PresignClient` type as the `PresignURL` interface defined in this middleware?  That provides stronger type association and reduces the need for empty interfaces.